# GBA 6070 - Programming Foundation for Business Analytics
# Dr. Mohammad Salehan
# Assignment 10 - Dataframe I|
Enter your name below.

In this assignment you will work with a weather dataset. Let's start with loading the dataset.

In [285]:
import pandas as pd
weather = pd.read_csv('weather2.csv')

We also use another dataset that contains information about each weather station.

In [286]:
stations = pd.read_excel('Station Information.xlsx')

1. Examine the shape of the both dataframes.

In [287]:
print(weather.shape)
stations.shape

(100000, 14)


(29745, 10)

2. Check the top 5 rows of each dataframe to see what it looks like.

In [288]:
weather.head()

Unnamed: 0  station_wban    year  month   day  hour  temperature  \
0           0  841170-99999  2010.0    1.0  15.0  16.0        270.0   
1           1  841170-99999  2010.0    1.0  15.0  17.0        270.0   
2           2  841170-99999  2010.0    1.0  15.0  18.0        276.0   
3           3  841170-99999  2010.0    1.0  15.0  19.0        280.0   
4           4  841170-99999  2010.0    1.0  15.0  20.0        280.0   

   dew_point  pressure  wind_direction  wind_speed  sky_condition  \
0      210.0   -9999.0           300.0        51.0            7.0   
1      210.0   -9999.0           310.0        57.0            7.0   
2      208.0   10115.0           310.0        51.0            7.0   
3      220.0   -9999.0           320.0        62.0            7.0   
4      220.0   -9999.0           300.0        72.0            7.0   

   precipitation_1h  precipitation_6h  
0           -9999.0           -9999.0  
1           -9999.0           -9999.0  
2           -9999.0           -9999.0  
3           -9999.0           -9999.0  
4           -9999.0           -9999.0

In [289]:
stations.head()

USAF   WBAN STATION NAME CTRY ST CALL    LAT     LON  ELEV(M)     BEGIN  \
0  7018  99999   WXPOD 7018  NaN     NaN   0.00   0.000   7018.0  20110309   
1  7026  99999   WXPOD 7026   AF     NaN   0.00   0.000   7026.0  20120713   
2  7070  99999   WXPOD 7070   AF     NaN   0.00   0.000   7070.0  20140923   
3  8260  99999    WXPOD8270  NaN     NaN   0.00   0.000      0.0  20050101   
4  8268  99999    WXPOD8278   AF     NaN  32.95  65.567   1156.7  20100519   

        END  
0  20130730  
1  20170822  
2  20150926  
3  20100920  
4  20120323

3. Drop the first column (Unnamed) of weather, then display the top row.

In [290]:
weather.drop(weather.columns[0],axis=1,inplace=True)
weather.head(1)

station_wban    year  month   day  hour  temperature  dew_point  pressure  \
0  841170-99999  2010.0    1.0  15.0  16.0        270.0      210.0   -9999.0   

   wind_direction  wind_speed  sky_condition  precipitation_1h  \
0           300.0        51.0            7.0           -9999.0   

   precipitation_6h  
0           -9999.0

Below, we combine USAF and WBAN columns of stations dataframe to create a new column named station_wban that looks like the one in weather dataframe.

In [291]:
stations['station_wban'] = stations['USAF'].astype(str)+'-'+stations['WBAN'].astype(str)
stations.head()

USAF   WBAN STATION NAME CTRY ST CALL    LAT     LON  ELEV(M)     BEGIN  \
0  7018  99999   WXPOD 7018  NaN     NaN   0.00   0.000   7018.0  20110309   
1  7026  99999   WXPOD 7026   AF     NaN   0.00   0.000   7026.0  20120713   
2  7070  99999   WXPOD 7070   AF     NaN   0.00   0.000   7070.0  20140923   
3  8260  99999    WXPOD8270  NaN     NaN   0.00   0.000      0.0  20050101   
4  8268  99999    WXPOD8278   AF     NaN  32.95  65.567   1156.7  20100519   

        END station_wban  
0  20130730   7018-99999  
1  20170822   7026-99999  
2  20150926   7070-99999  
3  20100920   8260-99999  
4  20120323   8268-99999

4. Join the 2 dataframes to add station name, country (CTRY) and state (ST CALL) to weather dataframe. There is no need for the rest of the columns in stations. You should use the proper type of join so that the number of rows in weather dataframe stays the same. At the end of this part, the shape of weather should be (100000, 16).

In [292]:
weather=pd.merge(weather,stations,on='station_wban',how='left').drop(['USAF','WBAN','LAT','LON','ELEV(M)','BEGIN','END'],axis=1)
weather.shape

(100000, 16)

5. Show descriptive statistics for all numeric columns of weather dataframe.

In [293]:
weather.describe()

year          month            day           hour  \
count  100000.00000  100000.000000  100000.000000  100000.000000   
mean     2009.34663       6.751760      15.871240      11.483780   
std         5.49655       3.392016       8.831961       6.926196   
min      1999.00000       1.000000       1.000000       0.000000   
25%      2005.00000       4.000000       8.000000       5.000000   
50%      2009.00000       7.000000      16.000000      11.000000   
75%      2015.00000      10.000000      24.000000      17.000000   
max      2019.00000      12.000000      31.000000      23.000000   

         temperature     dew_point       pressure  wind_direction  \
count  100000.000000  100000.00000  100000.000000   100000.000000   
mean      136.975390      81.14740    5797.130830       23.626870   
std       361.774103     362.78918    8303.764083     1264.655425   
min     -9999.000000   -9999.00000   -9999.000000    -9999.000000   
25%        61.000000      13.00000   10037.000000       70.000000   
50%       156.000000      99.00000   10140.000000      180.000000   
75%       239.000000     178.00000   10189.000000      270.000000   
max      9999.000000    9999.00000   10449.000000      999.000000   

         wind_speed  sky_condition  precipitation_1h  precipitation_6h  
count  100000.00000  100000.000000      100000.00000     100000.000000  
mean       33.54303   -1839.430760       -4556.22822      -9511.991990  
std       110.06946    3878.528361        4980.57331       2154.103991  
min     -9999.00000   -9999.000000       -9999.00000      -9999.000000  
25%        20.00000       0.000000       -9999.00000      -9999.000000  
50%        31.00000       2.000000          -1.00000      -9999.000000  
75%        46.00000       7.000000           0.00000      -9999.000000  
max       999.00000       9.000000        2794.00000       1278.000000

6. In all numeric columns, values 9999, -9999, 999, and -999 represent missing values. Replace all of them with np.nan using <a href='https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.replace.html' target='blank'>.replace() </a> method of dataframe.

In [294]:
import numpy as np
weather.replace([9999, -9999, 999, -999],np.nan,inplace=True)

7. Run descriptive statistics again to make sure the the values are transformed correctly.

In [295]:
weather.describe()

year          month            day           hour  \
count  100000.00000  100000.000000  100000.000000  100000.000000   
mean     2009.34663       6.751760      15.871240      11.483780   
std         5.49655       3.392016       8.831961       6.926196   
min      1999.00000       1.000000       1.000000       0.000000   
25%      2005.00000       4.000000       8.000000       5.000000   
50%      2009.00000       7.000000      16.000000      11.000000   
75%      2015.00000      10.000000      24.000000      17.000000   
max      2019.00000      12.000000      31.000000      23.000000   

        temperature     dew_point      pressure  wind_direction    wind_speed  \
count  99711.000000  99698.000000  78317.000000    97501.000000  99937.000000   
mean     144.152882     87.462597  10162.052505      173.352971     34.144951   
std      110.650054    108.562320     71.561724      112.883343     23.637969   
min     -289.000000   -333.000000   9833.000000        0.000000      0.000000   
25%       61.000000     14.000000  10121.000000       70.000000     20.000000   
50%      156.000000     99.000000  10159.000000      180.000000     31.000000   
75%      239.000000    178.000000  10204.000000      270.000000     46.000000   
max      420.000000    280.000000  10449.000000      360.000000    175.000000   

       sky_condition  precipitation_1h  precipitation_6h  
count   81570.000000      54426.000000       4863.000000  
mean        4.149736          1.315621         15.559120  
std         3.341461         18.754512         50.480649  
min         0.000000         -1.000000         -1.000000  
25%         0.000000          0.000000          0.000000  
50%         4.000000          0.000000          0.000000  
75%         8.000000          0.000000          8.000000  
max         9.000000       2794.000000       1278.000000

8. All weather meaures are inflated with a factor of 10. For example, 289 for temperature means 28.9 degrees. Divide all temperature values by 10. Use .apply() method and lambda function. Other columns, such as day and hour should not be affected.

In [296]:
weather['temperature']=weather['temperature'].apply(lambda x:x/10)
weather.describe()


year          month            day           hour  \
count  100000.00000  100000.000000  100000.000000  100000.000000   
mean     2009.34663       6.751760      15.871240      11.483780   
std         5.49655       3.392016       8.831961       6.926196   
min      1999.00000       1.000000       1.000000       0.000000   
25%      2005.00000       4.000000       8.000000       5.000000   
50%      2009.00000       7.000000      16.000000      11.000000   
75%      2015.00000      10.000000      24.000000      17.000000   
max      2019.00000      12.000000      31.000000      23.000000   

        temperature     dew_point      pressure  wind_direction    wind_speed  \
count  99711.000000  99698.000000  78317.000000    97501.000000  99937.000000   
mean      14.415288     87.462597  10162.052505      173.352971     34.144951   
std       11.065005    108.562320     71.561724      112.883343     23.637969   
min      -28.900000   -333.000000   9833.000000        0.000000      0.000000   
25%        6.100000     14.000000  10121.000000       70.000000     20.000000   
50%       15.600000     99.000000  10159.000000      180.000000     31.000000   
75%       23.900000    178.000000  10204.000000      270.000000     46.000000   
max       42.000000    280.000000  10449.000000      360.000000    175.000000   

       sky_condition  precipitation_1h  precipitation_6h  
count   81570.000000      54426.000000       4863.000000  
mean        4.149736          1.315621         15.559120  
std         3.341461         18.754512         50.480649  
min         0.000000         -1.000000         -1.000000  
25%         0.000000          0.000000          0.000000  
50%         4.000000          0.000000          0.000000  
75%         8.000000          0.000000          8.000000  
max         9.000000       2794.000000       1278.000000

9. Using a copy of your dataframe, replace all missing values with mean of each column.

In [297]:
weather_df=weather.fillna(weather.mean())
weather_df.head()

/var/folders/ft/nr64ylkn0lqb43gh8m5vvy3h0000gn/T/ipykernel_8276/455616676.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  weather_df=weather.fillna(weather.mean())


station_wban    year  month   day  hour  temperature  dew_point  \
0  841170-99999  2010.0    1.0  15.0  16.0         27.0      210.0   
1  841170-99999  2010.0    1.0  15.0  17.0         27.0      210.0   
2  841170-99999  2010.0    1.0  15.0  18.0         27.6      208.0   
3  841170-99999  2010.0    1.0  15.0  19.0         28.0      220.0   
4  841170-99999  2010.0    1.0  15.0  20.0         28.0      220.0   

       pressure  wind_direction  wind_speed  sky_condition  precipitation_1h  \
0  10162.052505           300.0        51.0            7.0          1.315621   
1  10162.052505           310.0        57.0            7.0          1.315621   
2  10115.000000           310.0        51.0            7.0          1.315621   
3  10162.052505           320.0        62.0            7.0          1.315621   
4  10162.052505           300.0        72.0            7.0          1.315621   

   precipitation_6h      STATION NAME CTRY ST CALL  
0          15.55912  ELOY ALFARO INTL   EC    SEMT  
1          15.55912  ELOY ALFARO INTL   EC    SEMT  
2          15.55912  ELOY ALFARO INTL   EC    SEMT  
3          15.55912  ELOY ALFARO INTL   EC    SEMT  
4          15.55912  ELOY ALFARO INTL   EC    SEMT

10. Calculate mean temperature, max wind speed, and min dew point for each station for each year using the dataframe from the previous step.

In [298]:
weather_df.groupby(['STATION NAME','year']).aggregate({'temperature':np.mean,'wind_speed':max,'dew_point':'min'})

temperature  wind_speed   dew_point
STATION NAME                  year                                       
BANGOR INTL                   2006.0    -0.086576        97.0 -267.000000
BIRMINGHAM INTERNATIONAL AIRP 1999.0    28.673373        67.0  168.000000
                              2007.0    29.559903        99.0  100.000000
                              2008.0     8.727381        98.0 -178.000000
                              2011.0    16.524815       139.0  -72.000000
BUCKLEY AIR FORCE BASE        2005.0    16.670402        98.0  -90.000000
                              2009.0    15.676407       144.0 -120.000000
                              2016.0     6.432739        98.0 -223.000000
BURGAS                        2007.0     6.693777       150.0  -91.000000
BURLINGTON INTERNATIONAL AIRP 2000.0     0.862165       129.0 -250.000000
                              2008.0     3.005140       139.0 -244.000000
CAIRO INTL                    2003.0    16.276334        99.0  -95.000000
                              2006.0    12.415385        51.0   47.000000
                              2015.0    18.624924        88.0  -40.000000
CAMPBELL AAF AIRPORT          2009.0     5.929665        98.0 -155.000000
CATANIA/SIGONELLA             1999.0    12.272839       175.0  -30.000000
                              2006.0    22.144752        98.0  -90.000000
CITY OF COLORADO SPRINGS MUNI 2002.0     0.153400        98.0 -267.000000
DAVIS MONTHAN AFB             2003.0    20.194030        93.0 -160.000000
DES MOINES INTERNATIONAL AIRP 2002.0    -0.375000        98.0 -206.000000
ELLINGTON FIELD AIRPORT       2013.0    29.141351        99.0   87.462597
ELOY ALFARO INTL              2010.0    26.217236        87.0   87.462597
                              2015.0    25.593533        99.0   87.462597
                              2019.0    23.636676        99.0   87.462597
FORBES FLD                    2002.0     5.127315       123.0 -140.000000
GRISSOM AFB AIRPORT           2016.0     6.551405       175.0  -99.000000
HECTOR INTERNATIONAL AIRPORT  2016.0    12.650715       160.0 -172.000000
                              2018.0    15.750718       154.0  -89.000000
HILO INTERNATIONAL AIRPORT    2017.0    24.594244        93.0  122.000000
HONOLULU INTERNATIONAL AIRPOR 2015.0    22.723810       144.0   72.000000
LANGLEY AFB AIRPORT           2008.0    24.344397        88.0  116.000000
MANAS                         2004.0     7.250646       110.0 -200.000000
MCCHORD AFB AIRPORT           2005.0    19.185776        72.0   10.000000
                              2009.0    14.185795        98.0  -60.000000
                              2012.0     4.693801       134.0  -71.000000
MINOT INTERNATIONAL ARPT      2007.0    -3.680451       165.0 -333.000000
NAVAL STATION ROOSEVELT ROADS 2011.0    26.928214        99.0   87.462597
OFFUTT AFB AIRPORT            2005.0    23.189887        93.0   90.000000
                              2017.0    21.064949        98.0    0.000000
                              2019.0    23.557562        82.0   80.000000
PAYA LEBAR                    2005.0    29.219368        67.0  210.000000
PITT GREENVILLE               2000.0    21.237470       129.0  -60.000000
PITTSBURGH INTERNATIONAL AIRP 2000.0     8.413071       139.0  -94.000000
                              2006.0     4.635102        99.0  -78.000000
                              2018.0    22.914980       129.0   89.000000
PORTLAND INTERNATIONAL AIRPOR 2009.0    16.973607        99.0  -94.000000
                              2010.0    15.341126        99.0  -50.000000
PORTLAND/HILLSBORO            2004.0    17.671759        82.0   10.000000
RICHMOND INTERNATIONAL AIRPOR 2003.0    25.575140        82.0  117.000000
                              2011.0    10.868392       139.0 -106.000000
SAN DIEGO INTERNATIONAL AIRPO 2014.0    22.361448        77.0  -39.000000
SINGAPORE CHANGI INTL         2001.0    26.770476        87.0   87.462597
SOUDA                         2001.0    11

11. When imputing missing values, replace each missing values with the mean of the measure for the same station. Use the dataframe from step 8.

In [299]:
weather['temperature']=weather.groupby('STATION NAME')['temperature'].transform(lambda x:x.fillna(x.mean()))
weather['dew_point']=weather.groupby('STATION NAME')['dew_point'].transform(lambda x:x.fillna(x.mean()))
weather['pressure']=weather.groupby('STATION NAME')['pressure'].transform(lambda x:x.fillna(x.mean()))
weather['wind_direction']=weather.groupby('STATION NAME')['wind_direction'].transform(lambda x:x.fillna(x.mean()))
weather['wind_speed']=weather.groupby('STATION NAME')['wind_speed'].transform(lambda x:x.fillna(x.mean()))
weather['sky_condition']=weather.groupby('STATION NAME')['sky_condition'].transform(lambda x:x.fillna(x.mean()))
weather['precipitation_1h']=weather.groupby('STATION NAME')['precipitation_1h'].transform(lambda x:x.fillna(x.mean()))
weather['precipitation_6h']=weather.groupby('STATION NAME')['precipitation_6h'].transform(lambda x:x.fillna(x.mean()))                                    
                                        

In [300]:
weather.isnull().sum()

station_wban            0
year                    0
month                   0
day                     0
hour                    0
temperature             0
dew_point               0
pressure             7181
wind_direction          0
wind_speed              0
sky_condition        2868
precipitation_1h    22960
precipitation_6h     7181
STATION NAME            0
CTRY                    0
ST CALL                 0
dtype: int64

In [301]:
weather.shape

(100000, 16)